rkt/gc.go

```diff
@@ -50,87 +48,78 @@ func runGC(args []string) (exit int) {
 		return 1
 	}
 
-	cs, err := getContainers()
-	if err != nil {
-		fmt.Fprintf(os.Stderr, "Unable to get containers list: %v\n", err)
+	if err := renameExited(); err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to rename exited containers: %v\n", err)
 		return 1
 	}
-	for _, c := range cs {
-		cp := filepath.Join(containersDir(), c)
-		l, err := lock.TryExclusiveLock(cp)
-		if err != nil {
-			fmt.Fprintf(os.Stderr, "Unable to open lock, ignoring %q: %v\n", c, err)
-			continue
-		}
-
-		fmt.Printf("Moving container %q to garbage\n", c)
-		err = os.Rename(cp, filepath.Join(garbageDir(), c))
-		if err != nil {
-			fmt.Println(err)
-		}
-		l.Close()
-	}
 
 	// clean up anything old in the garbage dir
-	err = emptyGarbage(flagGracePeriod)
-	if err != nil {
-		fmt.Println(err)
+	if err := emptyGarbage(flagGracePeriod); err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to empty garbage: %v\n", err)
 		return 1
 	}
 
 	return
 }
 
-// getContainers returns a slice representing the containers in the given rocket directory
-func getContainers() ([]string, error) {
-	cdir := containersDir()
-	ls, err := ioutil.ReadDir(cdir)
-	if err != nil {
-		return nil, fmt.Errorf("cannot read containers directory: %v", err)
-	}
-	var cs []string
-	for _, dir := range ls {
-		if !dir.IsDir() {
-			fmt.Fprintf(os.Stderr, "Unrecognized file: %q, ignoring", dir)
-			continue
+// renameExited renames exited containers to the garbage directory
+func renameExited() error {
+	if err := walkContainers(includeContainersDir, func(uuid string, c *container, err error) error {
+		if err != nil {
```

Comment:
I appreciate the push for similarity with `filepath.WalkFunc`, but this is probably showing that the analogy stretches a little thin since all walkfuncs so far do this same handling. I say check the uuid in walkContainers and simplify the func signature to just `func(c *container)` (since uuid is contained within it)